### PR TITLE
[DOC] Improved scipy.cluster.hierarchy documentation

### PR DESCRIPTION
**Pull request by carrdelling**
Added an example of use, starting from raw data

### CODE REVIEW

**Comment by matthew-brett** [comment]
(Am sitting next to @carrdelling right now but) this way of showing clustering looks nice to me, but I'm not a cluster person.  @deads @pv @rgommers - looks like you're the main maintainers here - does this make sense to you?   If so, then @carrdelling would like to continue with other examples.

File: scipy/cluster/hierarchy.py [hunk 1]
```diff
@@ -268,6 +268,63 @@ def single(y):
...
     scipy.spatial.distance.pdist : pairwise distance metrics
 
+    Examples
+    --------
+    >>> from scipy.cluster.hierarchy import single, fcluster
+    >>> from scipy.spatial.distance import pdist
+
+    First we need a toy dataset to play with:
+
+    x x    x x
```
Comment (ilayn) [inline scipy/cluster/hierarchy.py]: I don't know if these `x`s are meant to provide ASCII art but if so they need to be indented. Currently they don't render correctly as codeblocks

See https://6690-1460385-gh.circle-artifacts.com/0/html-scipyorg/generated/scipy.cluster.hierarchy.single.html#scipy.cluster.hierarchy.single

File: scipy/cluster/hierarchy.py [hunk 2]
```diff
@@ -294,6 +351,65 @@ def complete(y):
...
     scipy.spatial.distance.pdist : pairwise distance metrics
 
+    Examples
+    --------
+    >>> from scipy.cluster.hierarchy import complete, fcluster
+    >>> from scipy.spatial.distance import pdist
+
+    First we need a toy dataset to play with:
+
+    x x    x x
```
Comment (ilayn) [inline scipy/cluster/hierarchy.py]: Same issue as above

File: scipy/cluster/hierarchy.py (hunk 1 above)
Comment (pv) [reply scipy/cluster/hierarchy.py]: ... and hte previous line needs to end with double colon `::`,
http://www.sphinx-doc.org/en/master/usage/restructuredtext/basics.html

**Comment by matthew-brett** [comment]
@carrdelling - I sent a PR to your branch, fixing the docstring formatting.  Sorry, I should have spotted this at the time:

https://github.com/carrdelling/scipy/pull/2

@pv, @ilayn - thanks for the formatting comments - does this look like a good approach for making docstring based examples (that don't need plots)?  It looks useful to me.

**Comment by carrdelling** [comment]
@matthew-brett  - PR merged (thanks!). With this fixed, if @pv and @ilayn are happy with the format I can continue documenting the rest of the functions in the clustering module.

**Comment by ilayn** [comment]
LGTM. https://6862-1460385-gh.circle-artifacts.com/0/html-scipyorg/generated/scipy.cluster.hierarchy.complete.html#scipy.cluster.hierarchy.complete

File: scipy/cluster/hierarchy.py [hunk 3]
```diff
@@ -1774,6 +2170,68 @@ def fcluster(Z, t, criterion='inconsistent', depth=2, R=None, monocrit=None):
...
+    linkage : for information about hierarchical clustering methods work.
+
+    Examples
+    --------
+    >>> from scipy.cluster.hierarchy import fcluster
+
+    All cluster linkage methods - e.g. :func:`scipy.cluster.hierarchy.linkage`
+    generate a linkage matrix ``Z`` as their output:
+
+    >>> Z
```
Comment (ilayn) [inline scipy/cluster/hierarchy.py]: Z is not defined

**Comment by ilayn** [comment]
other than the typo in fcluster looks good to me

https://6914-1460385-gh.circle-artifacts.com/0/html-scipyorg/cluster.hierarchy.html

**Comment by carrdelling** [comment]
Yes, I need to expand that example so the generation of the `Z` matrix is included - I'll fix that this afternoon with the next batch of examples for clustering methods ;)

**Comment by ilayn** [comment]
@carrdelling Are you finished with this? Looks good to me. It would be really nice if we can complete and get rid of the top of the list given in #7168 about the `cluster` module. 

No pressure just asking, please feel free decline.

**Comment by carrdelling** [comment]
@ilayn Yeah, I want to complete the whole `cluster` module before closing this PR, maybe this weekend -  I'll ping you anyway once it is ready.

**Comment by matthew-brett** [comment]
Excellent - thanks very much.

**Comment by ilayn** [comment]
Great thanks!

**Comment by deads** [comment]
Sorry for the delay. I was traveling without access to e-mail. I am glad
Joaquin is adding examples to the functions. I can take a look.

**Comment by carrdelling** [comment]
Ok, so now that the tests are passing I guess that this is it - all public functions in `cluster.hierarchy` now have examples (and "See also" references where necessary). @matthew-brett, @ilayn, @deads, please feel free to review and suggest changes if you see anything weird.

Other than that, this PR should be ready to merge :)

**Comment by deads** [comment]
Thank you for your hard work here. Do you have a link to the built
documentation?

**Comment by carrdelling** [comment]
No - I've been building it locally:
```
>> cd scipy
>> python setup.py build_sphinx
```

Which works for reviewing - except that it does not have the same CSS styles as the original one in the website :(

**Comment by WarrenWeckesser** [comment]
Great to see this work.  I won't be able to review this any time soon, but I have one quick comment.  In the docstrings, don't use the `:func:` annotation.  Just use single backticks around the function name. Sphinx will take care the of link to the function's documentation automatically.

**Comment by ilayn** [comment]
The resulting rendering https://7022-1460385-gh.circle-artifacts.com/0/html-scipyorg/cluster.hierarchy.html

This is really great, thanks for taking your time with this. I'll review them very soon.

**Review by ilayn — APPROVED** [review]
Apart from one tiny detail this looks good to me.

File: scipy/cluster/hierarchy.py [hunk 4]
```diff
@@ -1252,6 +1648,77 @@ def cophenet(Z, Y=None):
         :math:`ij` th entry is the cophenetic distance between
         original observations :math:`i` and :math:`j`.
 
+    See Also
+    --------
+    linkage: for a description of what a linkage matrix is.
+    squareform: transforming condensed matrices into square ones.
```
[the remark this answers is not in the document]
Comment (ilayn) [reply scipy/cluster/hierarchy.py]: https://7022-1460385-gh.circle-artifacts.com/0/html-scipyorg/generated/scipy.cluster.hierarchy.cophenet.html#scipy.cluster.hierarchy.cophenet

**Comment by carrdelling** [comment]
Hey, thank you both for your comments!

I have added both @WarrenWeckesser and @ilayn suggestions.

However, the Travis build seems to be failing after these small updates - annoyingly enough, as the logs are just showing errors for tests on other modules. Do you have any idea about this? If not I can just try to roll back the last commits to see if that fixes the problem.

**Comment by rgommers** [comment]
Don't worry about the CI failures, those are unrelated to what you did. They will be addressed in gh-8887

**Comment by ilayn** [comment]
Circle CI has passed and I couldn't detect any rendering issues. https://7092-1460385-gh.circle-artifacts.com/0/html-scipyorg/cluster.hierarchy.html But we can fix it if anything is missed later, hence,  merging. 

Thanks a lot @carrdelling  and others. We got rid of one more module with this.

**Comment by rgommers** [comment]
Awesome, thanks @carrdelling, all!

**Comment by deads** [comment]
Looks good to me. Thanks @carrdelling :)

**Comment by carrdelling** [comment]
Glad to help ;) Thanks everyone!

**Pull request merged**